pandas/io/pytables.py

```diff
@@ -4959,11 +4959,12 @@ def _dtype_to_kind(dtype_str: str) -> str:
     return kind
 
 
-def _get_data_and_dtype_name(data: Union[np.ndarray, ABCExtensionArray]):
+def _get_data_and_dtype_name(data: ArrayLike):
     """
     Convert the passed data into a storable form and a dtype string.
     """
     if is_categorical_dtype(data.dtype):
+        assert isinstance(data, Categorical)  # for mypy
```

Comment:
I _think_ could just do this instead of the is_categorical_dtype check.

Reply:
seems reasonable

Reply:
Is there a way to avoid such additional isinstance checks for mypy?

Reply:
Not at the moment; this is oustanding on the mypy github:

https://github.com/python/mypy/issues/5206

Reply:
Isn't `typing.cast(..)` then more idiomatic for typing purposes?

Reply:
for "inclusion" of a type, cast provides the required semantics. for "exclusion" assert is more idomatic since cast can not provide type narrowing. so, for example `assert foo is not None` is the recommended approach for optional types, see https://mypy.readthedocs.io/en/stable/kinds_of_types.html#optional-types-and-the-none-type.

for casting, cast is static and can mask errors, assert isinstance is dynamic and can cause unintended runtime errors.

in the contributing guide, we have "...the use of cast is strongly discouraged". To avoid repeating this discussion on numerous pull requests. This discussion should be an independent issue and the contributing guide updated to reflect the consensus agreement on best practice/ more idiomatic approach.

for this PR. it shouldn't be a blocker.

Reply:
> in the contributing guide, we have "...the use of cast is strongly discouraged". 

Yes, but it is exactly a use case as the above (custom checker function that mypy doesn't understand) that it mentions is still OK if unavoidable 
(the contributing docs don't say to use `assert isinstance(..)` instead, at the moment)

Reply:
OK, but in the actual code already has been changed to `if isinstance(..)`, so ignore my further comments here

Reply:
the style guide goes on to say "Where applicable a refactor of the code to appease static analysis is preferable" and "... every effort should be exhausted to avoid cast..."

Reply:
I'm not against cast. There are pros and cons to both approaches. at the moment, though, when using asserts, we have an unwritten rule to add # for mypy comment to keep track of these until such time as we have a better/consensus solution. So I therefore don't see this as a blocker or discussion point in any PR.

Reply:
Only just seen https://github.com/pandas-dev/pandas/pull/32730#discussion_r393924535, didn't refresh window before commenting. 

@jorisvandenbossche we should open an issue to continue this discussion.

Reply:
> @jorisvandenbossche we should open an issue to continue this discussion.

#32785